### PR TITLE
Fix handling file url of file attribute values

### LINKED ISSUE #10831: There's inconsistency in a way we treat storage-based assets
## It looks that
 * `MEDIA_ROOT` and `MEDIA_URL` are settings, that Django’s core `FileSystemStorage` relies on (see: https://github.com/django/django/blob/main/django/core/files/storage.py#L212 )
 * in the same way exist `STATIC_ROOT` and `STATIC_URL` and Django’s core `StaticFilesStorage` ( see: https://github.com/django/django/blob/main/django/contrib/staticfiles/storage.py#L16)
 * neither `S3Boto3Storage` (see: https://github.com/jschneier/django-storages/blob/master/storages/backends/s3boto3.py#L89), nor `GoogleCloudStorage` (see: https://github.com/jschneier/django-storages/blob/master/storages/backends/gcloud.py#L104) relies on settings mentioned above, instead of that they introduced their own set of settings
 * if we want to use this `django-storages` ’s backends directly we cannot relies on at least `STATIC_URL` as even Django template tag doesn’t relies on it, it uses `staticfiles_storage.url(path)` instead (see: https://github.com/django/django/blob/main/django/templatetags/static.py#L129), and that required that second pull request to Saleor@3.5

## My suggestion
* treat `MEDIA_URL`, `MEDIA_ROOT`, `STATIC_URL` and `STATIC_ROOT` as not a way to go in Django, regardless of that storage you’re using, and rather as `FileSystemStorage`  and `StaticFilesStorage`  -specific settings
 * use `django-storages` backend-specific settings, when these storages in use
 * relies on
```python
from django.core.files.storage import default_storage
default_storage.url(path)
```
and
```python
from django.contrib.staticfiles.storage import staticfiles_storage
staticfiles_storage.url(path)
```
when URL is required.
instead of naive
```python
urljoin(settings.MEDIA_URL, path)
```

## Non-functional requirements
- [ ]  clean and document storage-oriented settings in `saleor/settings.py`

### PR DESCRIPTION
**Pull request by IKarbowiak**
Update `file_url` handling in values of file attributes
- reject URLs that do not start from the URLs root
- URL root is taken from the current storage
- use storage when resolving the URL

Resolves #10831

# Impact

* [ ] New migrations
* [ ] New/Updated API fields or mutations
* [ ] Deprecated API fields or mutations
* [ ] Removed API types, fields, or mutations
* [ ] Documentation needs to be updated

# Pull Request Checklist

* [ ] Privileged queries and mutations are guarded by proper permission checks
* [ ] Database queries are optimized and the number of queries is constant
* [ ] Database migration files are up to date
* [ ] The changes are tested
* [ ] GraphQL schema and type definitions are up to date
* [ ] Changes are mentioned in the changelog

### CODE REVIEW

File: saleor/graphql/core/types/common.py [hunk 1]
```diff
@@ -435,7 +437,10 @@ class File(graphene.ObjectType):
 
     @staticmethod
     def resolve_url(root, info):
-        return info.context.build_absolute_uri(urljoin(settings.MEDIA_URL, root.url))
+        # check if URL is absolute:
+        if urlparse(root.url).netloc:
```
Comment (magul) [inline saleor/graphql/core/types/common.py:441]: Do we anticipate a situation, that `File` or `Image` object will have absolute URL? (I recall discussion with @patrys that we will have `UrlField` at some point, but that will not touch `File` and `Image` at all?)

File: saleor/graphql/attribute/utils.py [hunk 2]
```diff
@@ -234,10 +236,18 @@ def clean_input(
         return cleaned_input
 
     @staticmethod
-    def _clean_file_url(file_url: Optional[str]):
+    def _clean_file_url(file_url: Optional[str], error_class):
         # extract storage path from file URL
+        storage_root_url = build_absolute_uri(default_storage.base_url)
```
Comment (magul) [inline saleor/graphql/attribute/utils.py:241]: there's not `base_url` attribute in all storage classes that we use.

File: saleor/graphql/attribute/utils.py [hunk 3]
```diff
@@ -234,10 +236,18 @@ def clean_input(
...
+        if file_url is not None and not file_url.startswith(
+            storage_root_url  # type: ignore
+        ):
+            raise ValidationError(
+                "The file_url must be the path to the default storage.",
+                code=error_class.INVALID.value,
+            )
         return (
-            re.sub(f"^{settings.MEDIA_URL}", "", urlparse(file_url).path)
+            re.sub(f"^{default_storage.base_url}", "", urlparse(file_url).path)
```
Comment (magul) [inline saleor/graphql/attribute/utils.py:250]: two issues here:
 * `base_url` doesn't exist in all storage classes
 * in `FileSystemStorage` it tend to return `settings.MEDIA_URL`, which is just a path prefix (so will not catch protocol and host part in absolute URL)

File: saleor/graphql/attribute/utils.py [hunk 4]
```diff
@@ -234,10 +236,18 @@ def clean_input(
...
             if file_url is not None
```
Comment (magul) [inline saleor/graphql/attribute/utils.py:251]: this `if` doesn't make sense, as few lines above you `raise ValidationError` in such situation.

**Review by magul — CHANGES_REQUESTED** [review]
`base_url` is definitely something that need be fixed. Also having unit tests, that actually test this logic using storage classes beyond `FileSystemStorage` would ensure, that it's going to work on AWS, GCP or Azure.

File: saleor/graphql/attribute/utils.py (hunk 4 above)
Comment (IKarbowiak) [reply saleor/graphql/attribute/utils.py:251]: No, I'm not raising an error for such a situation.

File: saleor/graphql/core/types/common.py (hunk 1 above)
Comment (IKarbowiak) [reply saleor/graphql/core/types/common.py:441]: For swatch attributes, we'll have. And we don't want to change anything for swatches.

File: saleor/graphql/attribute/utils.py (hunk 3 above)
Comment (IKarbowiak) [reply saleor/graphql/attribute/utils.py:250]: Yeah, but here I replacing the whole part of the protocol and host.

**Comment by IKarbowiak** [comment]
New PR: https://github.com/saleor/saleor/pull/10958

**Pull request closed without merging**